src/iocp/mod.rs

```diff
@@ -0,0 +1,821 @@
+            }
+        }
+    }
+
+    /// This socket state was notified; see if we need to update it.
+    fn feed_event(self: Pin<Arc<Self>>, poller: &Poller) -> io::Result<FeedEventResult> {
+        let inner = self.as_ref().data().project_ref();
+
+        let (afd_info, socket) = match inner {
+            PacketInnerProj::Socket { packet, socket } => (packet, socket),
+            PacketInnerProj::Wakeup { .. } => {
+                // The poller was notified.
+                return Ok(FeedEventResult::Notified);
+            }
+        };
+
+        let mut socket_state = socket.lock().unwrap_or_else(|e| e.into_inner());
+        let mut event = Event::none(socket_state.interest.key);
+
+        // Put ourselves into the idle state.
+        socket_state.status = SocketStatus::Idle;
+
+        // If we are waiting to be deleted, just return and let the drop handler do their thing.
+        if socket_state.waiting_on_delete {
+            return Ok(FeedEventResult::NoEvent);
+        }
+
+        unsafe {
+            // SAFETY: The packet is not in transit.
+            let iosb = &mut *self.as_ref().iosb().get();
+
+            // Check the status.
+            match iosb.Anonymous.Status {
+                STATUS_CANCELLED => {
+                    // Poll request was cancelled.
+                }
+
+                status if status < 0 => {
+                    // There was an error, so we signal both ends.
+                    event.readable = true;
+                    event.writable = true;
+                }
+
+                _ => {
+                    // Check in on the AFD data.
+                    let afd_data = &*afd_info.get();
+
+                    if afd_data.handle_count() >= 1 {
+                        let events = afd_data.events();
+
+                        // If we closed the socket, remove it from being polled.
+                        if events.contains(AfdPollMask::LOCAL_CLOSE) {
+                            let source = {
+                                let mut sources =
+                                    poller.sources.write().unwrap_or_else(|e| e.into_inner());
+                                sources.remove(&socket_state.socket).unwrap()
+                            };
+                            return source.begin_delete().map(|()| FeedEventResult::NoEvent);
+                        }
+
+                        // Report socket-related events.
+                        let (readable, writable) = afd_mask_to_event(events);
+                        event.readable = readable;
+                        event.writable = writable;
+                    }
+                }
+            }
+        }
+
+        // Filter out events that the user didn't ask for.
+        {
+            if !socket_state.interest.readable {
+                event.readable = false;
+            }
+
+            if !socket_state.interest.writable {
+                event.writable = false;
+            }
+        }
```

Comment:
```suggestion
    event.readable &= socket_state.interest.readable;
    event.writable &= socket_state.interest.writable;
```
would imo be easier to comprehend